<a href="https://colab.research.google.com/github/LeeSeungwon89/Machine-learning_Theory/blob/master/CHAPTER9%20%EC%B6%94%EC%B2%9C%20%EC%8B%9C%EC%8A%A4%ED%85%9C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. 추천 시스템의 개요와 배경**

## **1.1. 추천 시스템의 개요**

본서를 참고하시기 바랍니다.

## **1.2. 온라인 스토어의 필수 요소, 추천 시스템**

추천 시스템은 온라인에서 매우 필수적인 기술입니다. 전자상거래, 콘텐츠 제공 등 다양한 온라인 업종에서 사용됩니다. 

아래 데이터는 추천 시스템을 구성하는 데 사용될 수 있습니다.

- 사용자가 무엇을 클릭했습니까?

- 사용자가 어떤 상품을 확인하거나 장바구니에 넣었습니까?

- 사용자가 어떤 상품을 구매했습니까?

- 사용자가 제품과 영화를 어떻게 평가했습니까?

- 사용자가 본인의 취향을 어떻게 정의했습니까?

이 외에도 여러 데이터가 사용될 수 있습니다.

## **1.3. 추천 시스템의 유형**

추천 시스템의 유형은 바로 아래 파트에서 다룹니다. **콘텐츠 기반 필터링**, **협업 필터링**(**최근접 이웃 협업 필터링**, **잠재 요인 협업 필터링**)이 있습니다. 대부분의 온라인 스토어에서는 잠재 요인 협업 필터링을 사용하지만 최근에는 개인화 특성을 더 강화할 목적으로 하이브리드 형식(콘텐츠 기반 필터링 + 협업 기반 필터링)을 사용하는 경우도 많아지는 추세입니다.

기실 **4.3. 확률적 경사 하강법을 이용한 행렬 분해** 파트까지는 아래 정리본보다 본서와 [링크1](https://hipster4020.tistory.com/109), [링크2](https://big-dream-world.tistory.com/61?category=470634)에 더 상세하게 설명되어 있습니다. 공부 흔적을 남길 목적이므로 핵심 내용만 정리하겠습니다.

# **2. 콘텐츠 기반 필터링 추천 시스템**

**콘텐츠 기반 필터링(Content Based Filtering)**은 사용자가 특정 아이템을 선호하는 경우에 비슷한 콘텐츠를 가진 다른 아이템을 추천하는 방식입니다. 예컨대 사용자가 특정 영화에 높은 평점을 줬다면 그 영화의 장르, 배우, 감독, 영화 키워드 등 콘텐츠와 유사한 영화를 추천합니다. 

추천 시스템의 초창기에 주로 사용된 방식이지만 아이템 특성에 따라 여전히 많이 사용됩니다. 

# **3. 최근접 이웃 협업 필터링**

먼저 **협업 필터링(Collaborative Filtering)**에 대해 살펴볼 필요가 있습니다. 이 방식은 사용자 행동 양식(User Behavior)만을 토대로 추천합니다. 사용자 행동 양식이란 사용자가 아이템에 부여한 평점이나 상품 구매 이력 등을 예로 들 수 있습니다.

협업 필터링은 **사용자-아이템 평점 매트릭스(행렬)**처럼 축적된 사용자 행동 데이터를 토대로 사용자가 아직 평가하지 않은 아이템을 예측 평가(Predicted Rating) 하는 것이 주목표입니다. 사용자-아이템 평점 행렬은 많은 아이템을 열로 가진 다차원 행렬이며, 사용자가 평점을 매기는 경우가 많지 않으므로 희소 행렬입니다. 예컨대 사용자(행, Row)가 아이템 1 ~ 4(열, Column) 중에 아이템 1과 아이템 3만 평점을 부여했다면 이 평점들을 토대로 아이템 4의 평점을 예측 평가 합니다.

**최근접 이웃 협업 필터링(Nearest Neighbor Collaborative Filtering)**은 **메모리(Memory) 협업 필터링**으로도 불립니다. 여기서 아래 두 기반으로 나뉩니다.

- **사용자 기반(User-User)**: 당신과 비슷한 고객들이 다음 상품도 선택했습니다.

- **아이템 기반(Item-Item)**: 이 상품을 선택한 다른 고객들은 다음 상품도 선택했습니다.

첫 번째로, 사용자 기반 방식은 특정 사용자와 비슷한 다른 사용자를 TOP-N으로 선정하고, 이 TOP-N 사용자가 좋아하는 아이템을 추천합니다. 특정 사용자와 다른 사용자 간 유사도(Similarity)를 측정하여 가장 높은 유사도를 가진 TOP-N 사용자를 추출하고 선호하는 아이템을 추천하는 것입니다. 예컨대 사용자 1 ~ 3 중에 사용자 1에게 아이템을 추천하는 경우 사용자 3보다 사용자 2가 사용자 1과 비슷하게 평점을 부여했다면, 사용자 2가 사용한 아이템을 사용자 1에게 추천합니다.

참고로 유사도 측정에는 텍스트 분석에 사용되는 코사인 유사도가 가장 많이 사용됩니다. 추천 시스템에 사용되는 데이터 또한 텍스트 데이터처럼 희소 행렬이기 때문입니다.

두 번째로, 아이템 기반 방식은 아이템이 가지는 속성과는 무관하게 사용자가 특정 아이템을 선호하는지 여부(평점을 어떻게 부여했는지)가 추천 기준입니다. 사용자 기반 방식과 다르게 사용자-아이템 평점 행렬의 행과 열이 반대 형태입니다. 행이 아이템이고 열이 사용자인 것입니다. 예컨대 여러 사용자가 부여한 아이템 1 ~ 3의 평점 중에 아이템 1과 2의 평점이 아이템 3의 평점보다 유사하다면 사용자에게 아이템 2를 추천합니다.

대개 두 방식 중에 아이템 기반 방식이 더 높은 정확도를 보이며 대부분 활용됩니다. 사용자 기반 방식은 거의 사용되지 않습니다. 비슷한 아이템을 선호하거나 선택한다고 모든 사용자의 취향이 비슷하다고 단정하기엔 근거가 미약합니다. 유행을 타거나 유명한 아이템의 경우 취향과 무관하게 많은 사용자가 선택하기도 합니다. 더욱이 사용자들이 평점을 부여한 아이템의 경우 대체적으로 평점을 부여한 빈도가 많지 않은데 이를 토대로 유사도를 비교하는 것은 다소 무리가 있습니다.

최근접 이웃 협업 필터링은 콘텐츠 기반 필터링처럼 추천 시스템의 초창기에 주로 사용된 방식이지만 아이템 특성에 따라 여전히 많이 사용됩니다. 특히 아마존은 아이템 기반의 최근접 이웃 협업 필터링을 추천 엔진으로 사용합니다.

# **4. 잠재 요인 협업 필터링**

## **4.1. 잠재 요인 협업 필터링의 이해**

**잠재 요인 협업 필터링(Latent Factor Collaborative Filtering)**은 사용자-아이템 평점 행렬에 있는 잠재 요인을 추출하여 추천 예측을 수행하는 기법입니다. 잠재 요인은 **행렬 분해(Matrix Factorization)** 기법을 활용하여 대규모 다차원 행렬을 **SVD(Singular Value Decomposition, 특이값 분해)** 같은 차원 감소 기법으로 분해하는 과정에서 추출합니다. 즉 잠재 요인 협업 필터링은 행렬 분해 기법을 기반으로 합니다.

잠재 요인을 명확하게 정의내릴 순 없지만 이 잠재 요인을 기반으로 사용자-아이템 평점 행렬을 저차원 밀집 행렬인 **사용자-잠재 요인 행렬**과 아이템-잠재 요인 매트릭스의 전치 행렬인 **잠재 요인-아이템 행렬**로 분해할 수 있습니다. 두 행렬을 내적하여 새로운 예측 사용자-아이템 평점 행렬 데이터를 생성하고 사용자가 평점을 부여하지 않은 아이템에 예측 평점을 부여하는 것이 잠재 요인 협력 필터링의 핵심입니다.

행렬 분해를 거쳐 추출되는 잠재 요인이 무엇을 의미하는지는 알 수 없습니다. 다만 영화 평점 기반의 사용자-아이템 평점 행렬이라면 장르별 특성 선호도 정도로 가정할 수 있습니다. 사용자-잠재 요인 행렬은 영화 장르 선호도, 아이템-잠재 요인 행렬은 장르별 특성값으로 정의할 수 있습니다. 본서와 [링크](https://ariz1623.tistory.com/228)에서 이 가정에 대하여 그림과 함께 메커니즘을 설명합니다.

## **4.2. 행렬 분해의 이해**

행렬 분해와 예측 평점 도출에 대한 원리는 본서와 [링크](https://big-dream-world.tistory.com/69)를 참고하시기 바랍니다.

참고로 행렬 분해는 주로 SVD를 이용하지만 Null 값이 존재하는 행렬에는 적용할 수 없습니다. 이런 경우 **SGD(Stochastic Gradient Descent, 확률적 경사 하강법)**나 **ALS(Alternating Least Squares, 교대 최소 제곱법)**를 사용하여 SVD를 수행합니다.

## **4.3. 확률적 경사 하강법을 이용한 행렬 분해**

확률적 경사 하강법을 이용한 행렬 분해의 원리는 본서와 [링크](https://big-dream-world.tistory.com/69)를 참고하시기 바랍니다.

일반적으로 사용자-아이템 평점 행렬은 행렬 분해를 위해 단순히 예측 오류값의 최소화와 학습 시 과적합을 피하기 위해 규제를 반영한 비용 함수를 적용합니다. L2 규제를 반영하여 실제 R 행렬 값과 예측 R 행렬 값의 차이를 최소화하는 방향성을 가지고 P 행렬과 Q 행렬에 적절한 값을 반복적으로 업데이트하면서 최적화된 예측 R 행렬을 구하는 방식이 SGD 기반 행렬 분해입니다.만족할 만한 오류값을 가질 때까지 작업을 반복하면서 P와 Q 값을 근사화합니다.

SGD를 토대로 행렬 분해를 수행해 보겠습니다. 분해할 원본 행렬 R을 행렬 P와 행렬 R로 분해하고 다시 $P \times Q.T$로 예측 행렬을 생성하겠습니다. 수행 과정이 다소 복잡해 보일 수 있으므로 가볍게 살피고 넘어가도 좋습니다. 

In [ ]:
import numpy as np

# 원본 행렬 R을 생성합니다. 
R = np.array([[4, np.NaN, np.NaN, 2, np.NaN ],
              [np.NaN, 5, np.NaN, 3, 1 ],
              [np.NaN, np.NaN, 3, 4, 4 ],
              [5, 2, 1, 2, np.NaN ]])
# 행은 사용자, 열은 아이템으로 지정합니다.
num_users, num_items = R.shape
# 잠재 요인의 차원 K는 3으로 설정합니다. 
K=3
# 분해 행렬인 P와 Q의 크기를 지정하고 정규분포를 가진 무작위 값으로 부여합니다. 
np.random.seed(1)
P = np.random.normal(scale=1./K, size=(num_users, K))
Q = np.random.normal(scale=1./K, size=(num_items, K))

원본 행렬 R과 예측 행렬간 오차를 도출하는 새 함수를 선언하겠습니다. 이 함수는 원본 행렬 R에서 Null 값이 아닌 행렬 값의 인덱스를 추출하여 이 인덱스에 있는 행렬 R의 값과 분해된 행렬 P와 Q를 이용하여 다시 조합된 예측 행렬 값의 RMSE 값을 반환합니다.

In [ ]:
from sklearn.metrics import mean_squared_error

def get_rmse(R, P, Q, non_zeros):
    error = 0
    # 분해된 행렬 P와 Q.T를 내적하여 예측 행렬 R을 도출합니다.
    full_pred_matrix = np.dot(P, Q.T)
    
    # 원본 행렬 R에서 Null 값이 아닌 값의 인덱스를 추출하여
    # 원본 행렬 R과 예측 행렬의 RMSE를 추출합니다.
    x_non_zero_index = [non_zero[0] for non_zero in non_zeros]
    y_non_zero_index = [non_zero[1] for non_zero in non_zeros]
    R_non_zeros = R[x_non_zero_index, y_non_zero_index]
    full_pred_matrix_non_zeros = full_pred_matrix[x_non_zero_index, y_non_zero_index]
      
    mse = mean_squared_error(R_non_zeros, full_pred_matrix_non_zeros)
    rmse = np.sqrt(mse)
    
    return rmse

SGD 기법을 기반으로 행렬 분해를 수행하겠습니다. 먼저 행렬 R에서 Null 값을 제외한 데이터의 행렬 인덱스를 추출합니다.

In [ ]:
# R > 0인 행 인덱스, 열 인덱스, 값을 `non_zeros` 리스트에 저장합니다. 
non_zeros = [(i, j, R[i, j]) for i in range(num_users) for j in range(num_items) if R[i, j] > 0]

# SGD를 반복해서 업데이트할 횟수를 지정합니다.
steps = 1000
# SGD의 학습률을 지정합니다.
learning_rate = 0.01
# L2 규제 계수를 지정합니다.
r_lambda = 0.01

# SGD 방식을 토대로 행렬 P와 행렬 Q를 계속 업데이트합니다. 
for step in range(steps):
    for i, j, r in non_zeros:
        # 실제 값과 예측 값의 차이인 오류 값을 구합니다.
        eij = r - np.dot(P[i, :], Q[j, :].T)
        # 규제를 반영한 SGD 업데이트를 적용합니다.
        P[i, :] = P[i, :] + learning_rate * (eij * Q[j, :] - r_lambda * P[i,:])
        Q[j, :] = Q[j, :] + learning_rate * (eij * P[i, :] - r_lambda * Q[j,:])

    rmse = get_rmse(R, P, Q, non_zeros)
    # 50회 반복할 때마다 오류 값을 출력합니다.
    if (step % 50) == 0:
        print("# iteration step:", step, "| rmse:", rmse)

# iteration step: 0 | rmse: 0.016410959588961445
# iteration step: 50 | rmse: 0.01637308966218993
# iteration step: 100 | rmse: 0.016333882336881784
# iteration step: 150 | rmse: 0.016293661274599174
# iteration step: 200 | rmse: 0.01625272572302336
# iteration step: 250 | rmse: 0.016211338984681307
# iteration step: 300 | rmse: 0.016169725523756753
# iteration step: 350 | rmse: 0.016128072491623933
# iteration step: 400 | rmse: 0.016086533303745463
# iteration step: 450 | rmse: 0.016045231961723017
# iteration step: 500 | rmse: 0.016004267423347165
# iteration step: 550 | rmse: 0.015963717671935662
# iteration step: 600 | rmse: 0.01592364333325902
# iteration step: 650 | rmse: 0.015884090797210332
# iteration step: 700 | rmse: 0.01584509485896164
# iteration step: 750 | rmse: 0.015806680922219037
# iteration step: 800 | rmse: 0.01576886681829548
# iteration step: 850 | rmse: 0.015731664296432853
# iteration step: 900 | rmse: 0.0156950802377121
# iteration step: 950 | rmse: 0.015659117

예측 행렬을 생성해 보겠습니다.

In [ ]:
pred_matrix = np.dot(P, Q.T)
print('원본 행렬:\n', R)
print()
print('예측 행렬:\n', np.round(pred_matrix, 3))

원본 행렬:
 [[ 4. nan nan  2. nan]
 [nan  5. nan  3.  1.]
 [nan nan  3.  4.  4.]
 [ 5.  2.  1.  2. nan]]

예측 행렬:
 [[3.991 1.171 1.266 2.    1.649]
 [6.288 4.978 0.896 2.983 1.003]
 [6.401 0.866 2.987 3.978 3.986]
 [4.969 2.005 1.006 2.013 1.258]]


두 행렬의 값은 큰 차이가 없습니다. Null 값은 예측값으로 채워졌습니다.

# **5. 콘텐츠 기반 필터링 실습 - TMDB 5000 영화 데이터 세트**

캐글의 [TMDB 5000 영화 데이터 세트](https://www.kaggle.com/tmdb/tmdb-movie-metadata)에서 파일을 다운로드하여 콘텐츠 기반 필터링을 실습해 보겠습니다. 

## **5.1. 장르 속성을 이용한 영화 콘텐츠 기반 필터링**

콘텐츠 기반 필터링은 사용자가 특정 영화를 감상하고 그 영화를 좋아했다면 비슷한 특성이나 속성, 구성 요소 등을 가진 다른 영화를 추천하는 방식입니다. 예컨대 액션 장르인 영화 A를 재미있게 봤다면 해당하는 장르 중에 높은 평점을 받은 다른 영화를 추천하거나 영화 A를 제작한 감독의 다른 영화를 추천합니다. 즉, 다양한 콘텐츠(장르, 감독, 배우, 평점, 키워드, 영화 설명)는 영화(아이템, 상품, 서비스 등) 간 유사성을 판단하는 기준입니다.

콘텐츠 기반 필터링 추천 시스템을 영화 장르 속성을 기반으로 생성해 보겠습니다. 장르 피처 값의 유사도를 비교하여 높은 평점을 가진 영화를 추천하는 시스템입니다.

## **5.2. 데이터 로딩 및 가공**

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive/')

Mounted at /content/gdrive/


In [ ]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

movies = pd.read_csv('/content/gdrive/MyDrive/tmdb_5000/tmdb_5000_movies.csv')
display(movies.head(3))
print(movies.shape)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466

(4803, 20)


콘텐츠 기반 필터링 추천 분석에 적합한 피처만 추출하여 새 데이터프레임으로 정리하겠습니다.

In [ ]:
movies_df = movies[['id', 'title', 'genres', 'vote_average', 'vote_count', 'popularity', 'keywords', 'overview']]
movies_df.head(3)

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   

                                              genres  vote_average  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...           7.2   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...           6.9   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...           6.3   

   vote_count  popularity                                           keywords  \
0       11800  150.437577  [{"id": 1463, "name": "culture clash"}, {"id":...   
1        4500  139.082615  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2        4466  107.376788  [{"id": 470, "name": "spy"}, {"id": 818, "name...   

                                            overview  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...

결측치를 확인해 보겠습니다.

In [ ]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            4803 non-null   int64  
 1   title         4803 non-null   object 
 2   genres        4803 non-null   object 
 3   vote_average  4803 non-null   float64
 4   vote_count    4803 non-null   int64  
 5   popularity    4803 non-null   float64
 6   keywords      4803 non-null   object 
 7   overview      4800 non-null   object 
dtypes: float64(2), int64(2), object(4)
memory usage: 300.3+ KB


In [ ]:
movies_df.isnull().sum()

id              0
title           0
genres          0
vote_average    0
vote_count      0
popularity      0
keywords        0
overview        3
dtype: int64

In [ ]:
movies_df[movies_df['overview'].isnull()]

id                                       title  \
2656  370980  Chiamatemi Francesco - Il Papa della gente   
4140  459488                 To Be Frank, Sinatra at 100   
4431  292539                                 Food Chains   

                                   genres  vote_average  vote_count  \
2656        [{"id": 18, "name": "Drama"}]           7.3          12   
4140  [{"id": 99, "name": "Documentary"}]           0.0           0   
4431  [{"id": 99, "name": "Documentary"}]           7.4           8   

      popularity                                           keywords overview  
2656    0.738646  [{"id": 717, "name": "pope"}, {"id": 5565, "na...      NaN  
4140    0.050625  [{"id": 6027, "name": "music"}, {"id": 225822,...      NaN  
4431    0.795698                                                 []      NaN

결측치는 'overview' 피처에 3개만 존재합니다. 크게 중요하지 않을 가능성이 높으므로 별도로 처리하지는 않겠습니다. 아울러 다른 피처들은 일견 결측치가 존재하지 않는 것으로 확인되지만 위 데이터프레임에서 행 4431의 'keywords'의 값은 '[]'이므로 무의미한 값이라고 볼 수 있습니다. 일단 별도로 처리하지 않고 넘어가겠습니다.

다만 피처 몇 가지를 살펴볼 필요는 있습니다. 피처 'genres'와 'keywords'는 한꺼번에 여러 값을 표현하기 위해 리스트에 딕셔너리(dict)가 포함된 형태로 값을 가집니다. 예컨대 'genres'에 여러 값이 포함된 이유는 특정 영화의 장르가 한 종류일 수도 있고 여러 종류일 수도 있기 때문입니다. 데이터 분석을 위해 적합한 형태로 가공해야 합니다.

우선 영화 하나만 값을 펼쳐서 확인해 보겠습니다.

In [ ]:
pd.set_option('max_colwidth', 1000)
movies_df[['title', 'genres', 'keywords']][:1]

title  \
0  Avatar   

                                                                                                                                   genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          keywords  
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]

'name'을 키(key)로 하여금 값을 추출할 수 있는 형태입니다. 다만 단순한 문자열이므로 이 형태에서는 추출할 수는 없습니다. 리스트 객체로 변형해야 원하는 값을 추출할 수 있습니다. `ast` 모듈의 `literal_eval()` 메서드를 사용하여 문자열을 리스트 객체로 변형하겠습니다.


In [ ]:
from ast import literal_eval

movies_df['genres'] = movies_df['genres'].apply(literal_eval)
movies_df['keywords'] = movies_df['keywords'].apply(literal_eval)

두 피처의 'name' 키를 사용하여 값만 추출하고 피처에 저장하겠습니다.

In [ ]:
movies_df['genres'] = movies_df['genres'].apply(lambda x: [y['name'] for y in x])
movies_df['keywords'] = movies_df['keywords'].apply(lambda x: [y['name'] for y in x])

In [ ]:
pd.set_option('max_colwidth', 1000)
movies_df[['genres', 'keywords']][:1]

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                                                                                                                                                                                                   keywords  
0  [culture clash, future, space war, space colony, society, space travel, futuristic, romance, space, alien, tribe, alien planet, cgi, marine, soldier, battle, love affair, anti war, power relations, mind and soul, 3d]

## **5.3. 장르 콘텐츠 유사도 측정**

'genres' 피처는 리스트에 여러 장르가 포함된 형태입니다. 장르별 유사도를 측정하는 가장 간단한 방법은 이 피처 값을 문자열로 변환하여 `CountVectorizer` 클래스로 피처 벡터화를 수행하고 도출된 행렬을 코사인 유사도로 비교하는 것입니다. 이 과정을 요약하면 아래와 같습니다.

1. 문자열로 변환한 'genres' 피처를 카운트 기반으로 피처 벡터화 하여 행렬로 바꿉니다.

1. 이 행렬을 코사인 유사도로 비교합니다. 이를 위해 데이터 세트의 레코드별로 다른 레코드와 장르에서 코사인 유사도 값을 가지는 객체를 생성합니다.

1. 장르 유사도가 높은 영화 중에 평점이 높은 순으로 추천합니다.

먼저 'genres' 피처를 문자열로 변환하고 피처 벡터화를 수행하겠습니다.


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

movies_df['genres_literal'] = movies_df['genres'].apply(lambda x: (' ').join(x))
count_vect = CountVectorizer(min_df=0, ngram_range=(1, 2))
genre_mat = count_vect.fit_transform(movies_df['genres_literal'])
print(genre_mat.shape)

(4803, 276)


이 행렬에 `cosine_similarity()` 메서드로 코사인 유사도를 계산하면 행별(레코드별) 유사도를 가지며, 다시 말하면 `movies_df` 데이터프레임의 레코드별 장르 유사도 값을 가지는 것입니다. 코사인 유사도를 계산하고 레코드를 3개만 출력해 보겠습니다.

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

genre_sim = cosine_similarity(genre_mat, genre_mat)
print(genre_sim.shape)
print(genre_sim[:3])

(4803, 4803)
[[1.         0.59628479 0.4472136  ... 0.         0.         0.        ]
 [0.59628479 1.         0.4        ... 0.         0.         0.        ]
 [0.4472136  0.4        1.         ... 0.         0.         0.        ]]


코사인 유사도 결과가 담긴 `genre_sim` 객체를 토대로 개별 레코드에 대해 가장 장르 유사도가 높은 순으로 다른 레코드를 추출해야 합니다. 이를 수행할 방법은 `genre_sim` 객체의 기준 레코드별로 비교 대상인 레코드의 유사도 값이 높은 순으로 정렬된 행렬의 인덱스 값을 추출하는 것입니다. 값이 높은 순으로 정렬된 비교 대상 레코드의 유사도 값을 추출하는 것이 아니라 비교 대상 레코드의 인덱스를 추출하는 것입니다. `argsort()[:, ::-1]` 형식으로 오름차순 정렬을 수행하여 이를 구현하겠습니다.

In [ ]:
genre_sim_sorted_index = genre_sim.argsort()[:, ::-1]
print(genre_sim_sorted_index)

[[   0 3494  813 ... 3038 3037 2401]
 [ 262    1  129 ... 3069 3067 2401]
 [   2 1740 1542 ... 3000 2999 2401]
 ...
 [4800 3809 1895 ... 2229 2230    0]
 [4802 1594 1596 ... 3204 3205    0]
 [4802 4710 4521 ... 3140 3141    0]]


첫 번째 행인 `[   0 3494  813 ... 3038 3037 2401]`에서 `0`은 자신인 레코드이고 `3494`는 가장 유사한 레코드인 3494번 레코드입니다. `813`은 그 다음으로 유사한 레코드인 813번 레코드입니다. 이렇게 도출한 레코드 인덱스를 이용하면 특정 레코드와 코사인 유사도가 높은 레코드를 추출할 수 있습니다.

## **5.4. 장르 콘텐츠 필터링을 이용한 영화 추천**

장르 유사도를 기반으로 영화를 추천하는 함수를 선언하겠습니다. 인자를 전달하여 함수를 실행하면 추천 영화 정보를 가진 데이터프레임을 반환합니다.

In [ ]:
# `df`는 영화 데이터프레임, `sorted_indexes`는 레코드별 장르 코사인 유사도 인덱스를 가지는 객체,
# `title_name`은 고객이 선정한 추천 기준인 영화 제목, `top_n`은 추천할 영화 건수입니다. 
def find_sim_movie(df, sorted_indexes, title_name, top_n=0):
    # `df` 인자(영화 데이터프레임)에서 'title' 피처에
    # `title_name` 인자(추천 기준 영화 제목)가 포함된 데이터프레임을 출력합니다.
    title_movie = df[df['title'] == title_name]

    # 'title_name' 인자(추천 기준 영화 제목)가 포함된 데이터프레임의 인덱스 객체를 `ndarray` 형태로 반환합니다.
    title_index = title_movie.index.values
    # `sorted_index` 인자(코사인 유사도 인덱스 개체)에서 유사도가 높은 순으로
    # `top_n` 인자(추천할 영화 개수)만큼 인덱스를 추출합니다.
    similar_indexes = sorted_indexes[title_index, :(top_n)]

    # 추출된 `top_n` 인자의 인덱스를 출력합니다. 2차원 데이터입니다.
    print(similar_indexes)
    # 데이터프레임에서 인덱스로 사용하기 위해 1차원 배열로 변경합니다.
    similar_indexes = similar_indexes.reshape(-1)

    return df.iloc[similar_indexes]

영화 'The Godfather'와 유사한 영화 10개를 추천해 보겠습니다.

In [ ]:
similar_movies = find_sim_movie(movies_df, genre_sim_sorted_index, 'The Godfather', 10)
similar_movies[['title', 'vote_average']]

[[2731 1243 3636 1946 2640 4065 1847 4217  883 3866]]


title  vote_average
2731                          The Godfather: Part II           8.3
1243                                    Mean Streets           7.2
3636                                   Light Sleeper           5.7
1946  The Bad Lieutenant: Port of Call - New Orleans           6.0
2640         Things to Do in Denver When You're Dead           6.7
4065                                      Mi America           0.0
1847                                      GoodFellas           8.2
4217                                            Kids           6.8
883                              Catch Me If You Can           7.7
3866                                     City of God           8.1

'The Godfather: Part II'가 최선으로 추천되었습니다. 'GoodFellas'도 유사한 영화입니다. 그러나 'Light Sleeper', 'Mi America', 'Kids' 등의 영화는 'The Godfather'를 선호하는 사용자에게 추천하기엔 적합하지 않은 영화라고 합니다. 평점이 낮거나 심각하게 낮은 경우도 보입니다. 개선될 필요가 있습니다.

이번에는 더 많은 후보군을 선정하여 평점에 따라 필터링해서 최종적으로 추천하는 방식으로 바꿔보겠습니다. 평점 정보를 나타내는 'vote_average' 피처를 적용하겠습니다. 이 피처는 여러 관객이 평가한 평점을 평균한 것이므로 평점을 부여한 관객 수가 많고 적음에 따라 왜곡 정도가 결정됩니다. 'vote_count' 피처를 확인하여 평균 평점에 대한 왜곡 정도를 살펴볼 필요가 있습니다. 오름차순으로 10개만 출력해 보겠습니다.

In [ ]:
movies_df[['title', 'vote_average', 'vote_count']].sort_values('vote_average', ascending=False)[:10]

title  vote_average  vote_count
3519          Stiff Upper Lips          10.0           1
4247     Me You and Five Bucks          10.0           2
4045     Dancer, Texas Pop. 81          10.0           1
4662            Little Big Top          10.0           1
3992                 Sardaarji           9.5           2
2386            One Man's Hero           9.3           2
2970        There Goes My Baby           8.5           2
1881  The Shawshank Redemption           8.5        8205
2796     The Prisoner of Zenda           8.4          11
3337             The Godfather           8.4        5893

평점은 높지만 평가 횟수는 적은 영화를 확인할 수 있습니다. 왜곡된 평점 데이터는 평점에 평가 횟수를 반영하여 보정할 수 있습니다. 영화 평점 사이트인 IMDB에서는 평가 횟수에 대한 가중치가 부여된 평점 방식(Weighted Rating)을 사용합니다. 가중 평점 공식은 아래와 같습니다.

$$Weighted Rating = (\frac{v}{v+m} * R) + (\frac{m}{v+m} * C)$$

- $v$: 개별 영화의 평점 부여 횟수('vote_count' 피처에 해당함)

- $m$: 평점을 부여하기 위한 최소 투표 횟수

- $R$: 개별 영화의 평균 평점('vote_average' 피처에 해당함)

- $C$: 전체 영화에 대한 평균 평점(=`movies_df['vote_average'].mean()`)

$m$은 평점 부여 횟수에 따른 가중치를 직접 조절하는 역할을 합니다. 값을 높이면 평점 부여 횟수가 많은 영화에 더 큰 가중 평점을 부여합니다.

먼저 변수 $m$과 $C$를 만들어 보겠습니다. 변수 $m$은 전체 투표 횟수에서 상위 60%에 해당하는 횟수로 저장하겠습니다.

In [ ]:
m = movies_df['vote_count'].quantile(0.6)
C = movies_df['vote_average'].mean()
print('m:', round(m, 3), 'C:', round(C, 3))

m: 370.2 C: 6.092


기존 평점을 가중 평점으로 변환하는 새 함수를 선언하여 'weghted_vote' 피처를 생성하겠습니다.

In [ ]:
m = movies_df['vote_count'].quantile(0.6)
C = movies_df['vote_average'].mean()

def weighted_vote_average(record):
    v = record['vote_count']
    R = record['vote_average']

    return (v/(v+m)) * R + (m/(m+v)) * C

In [ ]:
movies_df['weighted_vote'] = movies_df.apply(weighted_vote_average, axis=1)

가중 평점이 높은 상위 10개만큼 출력해 보겠습니다.

In [ ]:
movies_df[['title', 'vote_average', 'weighted_vote', 'vote_count']].sort_values(
    'weighted_vote', ascending=False)[:10]

title  vote_average  weighted_vote  vote_count
1881  The Shawshank Redemption           8.5       8.396052        8205
3337             The Godfather           8.4       8.263591        5893
662                 Fight Club           8.3       8.216455        9413
3232              Pulp Fiction           8.3       8.207102        8428
65             The Dark Knight           8.2       8.136930       12002
1818          Schindler's List           8.3       8.126069        4329
3865                  Whiplash           8.3       8.123248        4254
809               Forrest Gump           8.2       8.105954        7927
2294             Spirited Away           8.3       8.105867        3840
2731    The Godfather: Part II           8.3       8.079586        3338

마지막으로 가중 평점에 기반하여 추천하는 최종 함수를 선언하겠습니다. 위에서 선언했던 `find_sim_movie()`장르 유사성이 높은 영화를 `top_n` 인자의 2배만큼 후보군으로 선정하고 가중 평점이 높은 순으로 `top_n` 인자만큼 추출하는 함수입니다.

In [ ]:
def find_sim_movie(df, sorted_indexes, title_name, top_n=0):
    title_movie = df[df['title'] == title_name]
    title_index = title_movie.index.values

    similar_indexes = sorted_indexes[title_index, :(top_n * 2)]
    similar_indexes = similar_indexes.reshape(-1)
    # 기준 영화의 인덱스는 제외합니다.
    similar_indexes = similar_indexes[similar_indexes != title_index]

    return df.iloc[similar_indexes].sort_values('weighted_vote', ascending=False)[:top_n]

'The Godfather'와 유사한 영화를 추천해 보겠습니다.

In [ ]:
similar_movies = find_sim_movie(movies_df, genre_sim_sorted_index, 'The Godfather', 10)
similar_movies[['title', 'vote_average', 'weighted_vote']]

title  vote_average  weighted_vote
2731       The Godfather: Part II           8.3       8.079586
1847                   GoodFellas           8.2       7.976937
3866                  City of God           8.1       7.759693
1663  Once Upon a Time in America           8.2       7.657811
883           Catch Me If You Can           7.7       7.557097
281             American Gangster           7.4       7.141396
4041              This Is England           7.4       6.739664
1149              American Hustle           6.8       6.717525
1243                 Mean Streets           7.2       6.626569
2839                     Rounders           6.9       6.530427

본서에서는 이전보다 'The Godfather'와 유사한 영화를 추천했다고 명시하고 있습니다. 

다만 장르만으로는 영화의 요소나 분위기, 개인 취향을 반영하기엔 역부족입니다. 예컨대 선호하는 영화배우나 감독이 영화 결정 요인이 될 수도 있기 때문입니다. 따라서 콘텐츠 기반 필터링 추천 시스템은 다른 콘텐츠도 기반으로 삼아 고도화할 필요가 있습니다. 본서에서는 이 추천 시스템을 고도화하는 추가적인 과정을 다루고 있지 않습니다.

# **6. 아이템 기반 최근접 이웃 협업 필터링 실습**

사용자 기반과 아이템 기반 중 추천 정확도가 더 높은 아이템 기반으로 협업 힐터링을 구현해 보겠습니다. [Grouplens 사이트의 Movielens 데이터 세트](https://grouplens.org/datasets/movielens/latest/)를 사용하겠습니다.


## **6.1. 데이터 가공 및 변환**

데이터를 준비하겠습니다. 먼저 `movies.csv` 파일을 확인해 보겠습니다.

In [1]:
from google.colab import drive
drive.mount('/content/gdrive/')

Mounted at /content/gdrive/


In [2]:
import pandas as pd
import numpy as np

movies = pd.read_csv('/content/gdrive/MyDrive/ml-latest-small/ml-latest-small/movies.csv')
print(movies.shape)
movies.head()

(9742, 3)


movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [3]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9742 non-null   int64 
 1   title    9742 non-null   object
 2   genres   9742 non-null   object
dtypes: int64(1), object(2)
memory usage: 228.5+ KB


영화(아이템)와 관련된 데이터이며 결측치는 확인되지 않습니다.  피처 'genres'는 장르를 의미하며 `|`로 나뉩니다.

다음은 `ratings.csv` 파일을 확인해 보겠습니다.

In [4]:
ratings = pd.read_csv('/content/gdrive/MyDrive/ml-latest-small/ml-latest-small/ratings.csv')
print(ratings.shape)
ratings.head()

(100836, 4)


userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [5]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


관객(사용자)와 관련된 데이터이며 결측치는 확인되지 않습니다. 피처 4개 중에 'timestamp'는 분석 작업에 불필요한 피처로 사료됩니다.

협업 필터링은 `ratings` 데이터프레임처럼 사용자와 아이템 간 평점(또는 다른 유형의 액션)에 기반한 추천 시스템입니다. 이를 구현하려면 `ratings` 데이터프레임의 행을 사용자('userId' 피처)로 두고 열을 모든 영화('movieId'피처)로 둬야 합니다. 판다스의 `pivot_table()` 메서드를 적용하여 행 레벨을 열(피처) 레벨로 바꾸겠습니다.

In [7]:
ratings = ratings[['userId', 'movieId', 'rating']]
ratings_matrix = ratings.pivot_table('rating', index='userId', columns='movieId')
ratings_matrix.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     NaN     4.0     NaN     NaN     4.0     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  193583  193585  193587  193609  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  
4           NaN     NaN     NaN     NaN  
5           NaN     NaN     NaN     NaN  

[5 rows x 9724 columns]

영화명이 모두 숫자로 부여되어 있습니다. 'movieId'의 숫자 피처에 대응하는 영화명은 'movies' 데이터프레임에 정리되어 있습니다. 두 데이터프레임('movies'와 'ratings')을 조인하여 'movieId'에 대응하는 'title' 피처 값을 포함시키겠습니다.

In [11]:
rating_movies = pd.merge(ratings, movies, on='movieId')
rating_movies.head()

userId  movieId  rating             title  \
0       1        1     4.0  Toy Story (1995)   
1       5        1     4.0  Toy Story (1995)   
2       7        1     4.5  Toy Story (1995)   
3      15        1     2.5  Toy Story (1995)   
4      17        1     4.5  Toy Story (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1  Adventure|Animation|Children|Comedy|Fantasy  
2  Adventure|Animation|Children|Comedy|Fantasy  
3  Adventure|Animation|Children|Comedy|Fantasy  
4  Adventure|Animation|Children|Comedy|Fantasy

In [17]:
ratings_matrix = rating_movies.pivot_table('rating', index='userId', columns='title')
ratings_matrix.head()

title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1              NaN                                      NaN   
2              NaN                                      NaN   
3              NaN                                      NaN   
4              NaN                                      NaN   
5              NaN                                      NaN   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                          NaN                  NaN   
2                          NaN                  NaN   
3                          NaN                  NaN   
4                          NaN                  NaN   
5                          NaN                  NaN   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                             NaN                              NaN   
2                             NaN                              NaN   
3                             NaN                              NaN   
4                             NaN                              NaN   
5                             NaN                              NaN   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                      NaN                   NaN                          NaN   
2                      NaN                   NaN                          NaN   
3                      NaN                   NaN                          NaN   
4                      NaN                   NaN                          NaN   
5                      NaN                   NaN                          NaN   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                                  NaN  ...          NaN           NaN   
2                                  NaN  ...          NaN           NaN   
3                                  NaN  ...          NaN           NaN   
4                                  NaN  ...          NaN           NaN   
5                                  NaN  ...          NaN           NaN   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1                 NaN                      NaN   
2                 NaN                      NaN   
3                 NaN                      NaN   
4                 NaN                      NaN   
5                 NaN                      NaN   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                     NaN        
2                                                     NaN        
3                                                     NaN        
4                                                     NaN        
5                                                     NaN        

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1                   NaN         NaN                             NaN   
2                   NaN         NaN                             NaN   
3                   NaN         NaN                             NaN   
4                   NaN         NaN                             NaN   
5                   NaN         NaN                             NaN   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                         4.0                                        NaN  
2                         NaN                            

`NaN` 값이 많습니다. 사용자가 평점을 부여하지 않은 영화가 열로 변환되었기 때문입니다. 이 `Nan` 값은 0점을 의미하는 `0`으로 바꾸겠습니다.


In [18]:
ratings_matrix = ratings_matrix.fillna(0)
ratings_matrix.head()

title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1              0.0                                      0.0   
2              0.0                                      0.0   
3              0.0                                      0.0   
4              0.0                                      0.0   
5              0.0                                      0.0   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                          0.0                  0.0   
2                          0.0                  0.0   
3                          0.0                  0.0   
4                          0.0                  0.0   
5                          0.0                  0.0   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                             0.0                              0.0   
2                             0.0                              0.0   
3                             0.0                              0.0   
4                             0.0                              0.0   
5                             0.0                              0.0   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                      0.0                   0.0                          0.0   
2                      0.0                   0.0                          0.0   
3                      0.0                   0.0                          0.0   
4                      0.0                   0.0                          0.0   
5                      0.0                   0.0                          0.0   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                                  0.0  ...          0.0           0.0   
2                                  0.0  ...          0.0           0.0   
3                                  0.0  ...          0.0           0.0   
4                                  0.0  ...          0.0           0.0   
5                                  0.0  ...          0.0           0.0   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1                 0.0                      0.0   
2                 0.0                      0.0   
3                 0.0                      0.0   
4                 0.0                      0.0   
5                 0.0                      0.0   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                     0.0        
2                                                     0.0        
3                                                     0.0        
4                                                     0.0        
5                                                     0.0        

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1                   0.0         0.0                             0.0   
2                   0.0         0.0                             0.0   
3                   0.0         0.0                             0.0   
4                   0.0         0.0                             0.0   
5                   0.0         0.0                             0.0   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                         4.0                                        0.0  
2                         0.0                            

## **6.2. 영화 간 유사도 산출**



---

고민 중

---



## **6.3. 아이템 기반 최근접 이웃 협업 필터링으로 개인화된 영화 추천**

# **7. 행렬 분해를 이용한 잠재 요인 협업 필터링 실습**

# **8. 파이썬 추천 시스템 패키지 - Surprise**

## **8.1. Surprise 패키지 소개**

## **8.2. Surprise를 이용한 추천 시스템 구축**

## **8.3. Surprise 주요 모듈 소개**

### **8.3.1. Dataset**

### **8.3.2. OS 파일 데이터를 Surprise 데이터 세트로 로딩**

### **8.3.3. 판다스 DataFrame에서 Surprise 데이터 세트로 로딩**

## **8.4. Surprise 추천 알고리즘 클래스**

## **8.5. 베이스라인 평점**

## **8.6. 교차 검증과 하이퍼 파라미터 튜닝**

## **8.7. Surprise를 이용한 개인화 영화 추천 시스템 구축**

# **9. 정리**